magnet_coils.py

```diff
@@ -0,0 +1,350 @@
+            '[\'rectangle\' (str), width (float, cm), thickness (float, cm)]'
+        )
+        raise ValueError(
+            'Magnet cross-section must be either a circle or rectangle. The '
+            'first entry of the list defining the cross-section must be the '
+            'shape, with the following entries defining the shape parameters.\n'
+            '\n'
+            'For a circular cross-section, the list format is\n'
+            '[\'circle\' (str), radius (float, cm)]\n'
+            '\n'
+            'For a rectangular cross-section, the list format is\n'
+            '[\'rectangle\' (str), width (float, cm), thickness (float, cm)]'
+        )
+    
+    # Create cross-section for sweep
+    if shape == 'circle':
+        cubit.cmd(
+            f'create surface circle radius {radius} zplane'
+        )
+    else:
+        cubit.cmd(
+            f'create surface rectangle width {thickness} height {width} zplane'
+        )
+    # Store cross-section index
+    cs_id = cubit.get_last_id("surface")
+    # Cross-section initially populated with normal oriented along z
+    # axis
+    cs_axis = np.array([0, 0, 1])
+
+    # Initialize volume index storage list
+    vol_id = []
+
+    # Initialize path list
+    path = []
+
+    # Extract filament data
+    for filament in filaments:
+        
+        # Create vertices in filament path
+        for x, y, z in filament:
+            # Create vertex in path
+            cubit.cmd(f'create vertex {x} {y} {z}')
+            # Append vertex to path list
+            path += [cubit.get_last_id("vertex")]
+        
+        # Ensure final vertex in path is the same as the first
+        path += [path[0]]
+        
+        # Create spline for path
+        cubit.cmd(
+            f'create curve spline location vertex ' +
+            ' '.join(str(i) for i in path)
+        )
+        # Store curve index
+        curve_id = cubit.get_last_id("curve")
+
+        # Define new surface normal vector as that pointing between path
+        # points adjacent to initial point
+
+        # Extract path points adjacent to initial point
+        next_pt = np.array(cubit.vertex(path[1]).coordinates())
+        last_pt = np.array(cubit.vertex(path[-2]).coordinates())
+        # Compute tangent at path origin (direction in which to align surface
+        # normal)
+        tang = np.subtract(next_pt, last_pt)
+        tang = tang/np.linalg.norm(tang)
+        
+        # Define axis and angle of rotation to orient cross-section along
+        # defined normal
+
+        # Define axis of rotation as orthogonal to both z axis and surface
+        # normal
+        rot_axis = np.cross(cs_axis, tang)
+        rot_axis = rot_axis/np.linalg.norm(rot_axis)
+        # Compute angle by which to rotate cross-section to orient along
+        # defined surface normal
+        dot_prod = np.inner(cs_axis, tang)
+        rot_ang_norm = np.arccos(dot_prod)
```

Comment:
I think it's probably readable to combine these:
```suggestion
        rot_ang_norm = np.arccos(np.inner(cs_axis, tang))
```